File: Lintcode/8.Rotate_String.py
class Solution:
    # @param s: a list of char
    # @param offset: an integer 
    # @return: nothing
    def rotateString(self, s, offset):
	    # write you code here
	    if s is None or len(s)==0:
	        return
	    
	    offset %= len(s)
	    '''
	    #it's meaningless when @param s: a list of char
	    #while it's meaningful when @param s: a string
	    before = s[:len(s) - offset]
	    after = s[len(s)-offset:]
	    s = before[::-1] + after[::-1]
	    s = s[::-1]
	    '''
	    self.reverse(s,0,len(s)-offset-1)
	    self.reverse(s,len(s)-offset,len(s)-1)
	    self.reverse(s,0,len(s)-1)

    def reverse(self,temp,begin,end):
        while begin < end:
            temp[begin],temp[end] = temp[end],temp[begin]
            begin +=1
            end -= 1

File: Lintcode/8.Rotate_String.cpp
class Solution {
public:
    /**
     * @param str: a string
     * @param offset: an integer
     * @return: nothing
     */
    void rotateString(string &str,int offset){
        //wirte your code here
        
        if(str.size() == 0 || str.empty())
            return ;
            
        int offset_temp;
        offset_temp = offset % str.size();
        /* 
        string str_temp = "";
        if (offset_temp!=0)
        {
            str_temp = str.substr(str.size()-offset_temp,offset_temp) + str.substr(0,str.size()-offset_temp);
            str = str_temp;
        }
        */
        int length = str.size();
        if (offset_temp!=0)
        {
            reverse(str,0,length-1-offset_temp);
            reverse(str,length-offset_temp,length-1);
            reverse(str,0,length-1);
        }
    }
    
    void reverse(string &str,int begin,int end)
    {
        while (begin < end)
        {
            char temp = str[begin];
            str[begin] = str[end];
            str[end] = temp;
            begin ++;
            end --;
        }
    }
};
